components/brave_wallet/common/cardano_address_unittest.cc

```diff
@@ -0,0 +1,161 @@
+inline constexpr char kTestnetRewardScript[] =
+    "stake_test17rphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtcljw6kf";
+inline constexpr char kByronMainnetYoroi[] =
+    "Ae2tdPwUPEZFRbyhz3cpfC2CumGzNkFBN2L42rcUc2yjQpEkxDbkPodpMAi";
+inline constexpr char kByronTestnetDaedalus[] =
+    "37btjrVyb4KEB2STADSsj3MYSAdj52X5FrFWpw2r7Wmj2GDzXjFRsHWuZqrw7zSkwopv8Ci3VW"
+    "eg6bisU9dgJxW5hb2MZYeduNKbQJrqz3zVBsu9nT";
+}  // namespace
+
+// https://github.com/input-output-hk/cardano-js-sdk/blob/5bc90ee9f24d89db6ea4191d705e7383d52fef6a/packages/util-dev/src/Cip19TestVectors.ts#L31
+TEST(CardanoAddress, TestVectors) {
+  EXPECT_FALSE(CardanoAddress::FromString(""));
+
+  auto unsupported = std::to_array({
+      /*kBasePaymentKeyStakeKey,*/
+      kBasePaymentScriptStakeKey,
+      kBasePaymentKeyStakeScript,
+      kBasePaymentScriptStakeScript,
+      kPointerKey,
+      kPointerScript,
+      kEnterpriseKey,
+      kEnterpriseScript,
+      kRewardKey,
+      kRewardScript,
+      /*kTestnetBasePaymentKeyStakeKey,*/
+      kTestnetBasePaymentScriptStakeKey,
+      kTestnetBasePaymentKeyStakeScript,
+      kTestnetBasePaymentScriptStakeScript,
+      kTestnetPointerKey,
+      kTestnetPointerScript,
+      kTestnetEnterpriseKey,
+      kTestnetEnterpriseScript,
+      kTestnetRewardKey,
+      kTestnetRewardScript,
+      kByronMainnetYoroi,
+      kByronTestnetDaedalus,
+  });
+
+  for (auto* address : unsupported) {
+    EXPECT_FALSE(CardanoAddress::FromString(address))
+        << testing::Message(address);
+  }
+
+  auto base_payment_key_stake_key =
+      CardanoAddress::FromString(kBasePaymentKeyStakeKey);
+  ASSERT_TRUE(base_payment_key_stake_key);
+  EXPECT_FALSE(base_payment_key_stake_key->IsTestnet());
+  EXPECT_EQ(base_payment_key_stake_key->ToString(), kBasePaymentKeyStakeKey);
+  EXPECT_EQ(base::HexEncode(base_payment_key_stake_key->ToCborBytes()),
+            std::string("01") + kPaymentPartHash + kStakePartHash);
+  EXPECT_EQ(
+      base_payment_key_stake_key,
+      CardanoAddress::FromParts(false, test::HexToArray<28>(kPaymentPartHash),
+                                test::HexToArray<28>(kStakePartHash)));
+
+  auto testnet_base_payment_key_stake_key =
+      CardanoAddress::FromString(kTestnetBasePaymentKeyStakeKey);
+  ASSERT_TRUE(testnet_base_payment_key_stake_key);
+  EXPECT_TRUE(testnet_base_payment_key_stake_key->IsTestnet());
+  EXPECT_EQ(testnet_base_payment_key_stake_key->ToString(),
+            kTestnetBasePaymentKeyStakeKey);
+  EXPECT_EQ(base::HexEncode(testnet_base_payment_key_stake_key->ToCborBytes()),
+            std::string("00") + kPaymentPartHash + kStakePartHash);
+  EXPECT_EQ(
+      testnet_base_payment_key_stake_key,
+      CardanoAddress::FromParts(true, test::HexToArray<28>(kPaymentPartHash),
+                                test::HexToArray<28>(kStakePartHash)));
+}
+
+TEST(CardanoAddress, InvalidInput) {
+  auto invalid_cases = std::to_array({
+      "",
```

Comment:
It's very difficult to tell specifically what these test cases are checking. There are many places where `CardanoAddress` could return nullopt and I don't know which of those cases is being test by each of these. This makes it difficult as a reviewer to verify that we're testing all of those cases and also makes it difficult to know what is broken if one of them fails. It would be even better to split them out into individual test cases that clearly state what is being tested in each.